Chore: Report frontend metrics for JS/CSS load time

ifrost opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

Grafana backend stores the following frontend metrics in Prometheus data base [(frontendmetrics.go)](https://github.com/grafana/grafana/blob/main/pkg/infra/metrics/frontendmetrics.go#L39):
* `frontend_boot_load_time_seconds`
* `frontend_boot_first_paint_time_seconds`
* `frontend_boot_js_done_time_seconds`

However metrics that are sent from the frontend are:

<img width="632" alt="Screen Shot 2021-11-29 at 10 43 43" src="https://user-images.githubusercontent.com/745532/143844556-84bbafb7-ba39-41a6-8e45-2f16261e52b4.png">

Meaning:
* `frontend_boot_js_done_time_seconds` is not passed from the frontend
* first contentful paint is passed from the frontend but not used in the backend
* css load time (measured in [index-template.htm](https://github.com/grafana/grafana/blob/main/public/views/index-template.html#L29)) is not used

This PR adds:
* Storing additional FCP metric (`frontend_boot_first_contentful_paint_time_seconds`) 
* Passing `frontend_boot_js_done_time_seconds` and `frontend_boot_css_time_seconds`

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes N/A

**Special notes for your reviewer**:

_How to test it?_ Use devenv prometheus (`make run devenv sources=prometheus`) and query for metrics reported in `frontendmetrics.go`.

<img width="1480" alt="Screen Shot 2021-11-29 at 11 03 53" src="https://user-images.githubusercontent.com/745532/143847705-a0b75684-6d1a-48ab-841d-e10aed3452d7.png">

ifrost committed:
Report frontend metrics for js/css load time

ifrost committed:
Add FCP metric

aocenas commented on public/app/app.ts:
```diff
@@ -179,8 +182,9 @@ function initEchoSrv() {
 
-      const loadMetric = performance.getEntriesByName(loadMetricName)[0];
-      reportPerformance(loadMetric.name, Math.round(loadMetric.startTime + loadMetric.duration) / 1000);
+      reportMetricPerformance(loadMetricName);
```
What is this actually measuring? We mark it on line 174 and then report it here. It seems like it is only measuring how long it takes to report the paint metrics?

ifrost replied:
It measures the time up until that line is executed - from initial page load to window load event.

aocenas commented on public/app/app.ts:
```diff
@@ -166,6 +166,9 @@ function initEchoSrv() {
+    const jsLoadMetricName = 'frontend_boot_js_done_time_seconds';
+    const cssLoadMetricName = 'frontend_boot_css_time_seconds';
 
     if (performance && performance.getEntriesByType) {
```
This is bit confusing. To actually measure anything we have to explicitly report it but also backend needs to know about the measurement. So gathering it dynamically like this does not make sense to me. If browser starts reporting some more metrics we wouldn't automatically add it to prometheus because the histogram isn't registered. So I think it would make sense to just statically write down the metrics here for the paints and comment that it needs to match the `frontendmetrics.go` registered histograms.

ifrost replied:
It measures the time up until that line is executed - from initial page load to window load event.

ifrost replied:
Makes sense. I'll change them to be statically defined in frontend.

aocenas commented on public/app/app.ts:
```diff
@@ -166,6 +166,9 @@ function initEchoSrv() {
 
   window.addEventListener('load', (e) => {
     const loadMetricName = 'frontend_boot_load_time_seconds';
+    // should match names in public/views/index-template.html
```
nit: I would reword this to say these are 'marked' in public/views/index-template.html to keep with the `performance` API naming.

ifrost replied:
Updated.

aocenas reviewed (commented):
Left some comments which may not be really about the particular changes here but thought they are important as the code confused me quite a bit.

ifrost committed:
Change comment

ifrost committed:
Simplify reporting frontend metrics

ifrost committed:
Merge branch 'main' into ifrost/frontend-metrics

# Conflicts:
#	pkg/api/frontend_metrics.go

ifrost commented:
A couple of changes:

I'm not sure why we use `(metric.startTime + metric.duration)`. From docs it seems that `metric.startTime` should be the timestamp from [the time origin](https://developer.mozilla.org/en-US/docs/Web/API/DOMHighResTimeStamp#the_time_origin), which should be enough to report at what point in document lifecycle the event happened.

What is more `metric.duration` should always be `0`  in our cases because:
- `performance.mark()` entries have duration of 0 (https://developer.mozilla.org/en-US/docs/Web/API/PerformanceMark)
- `first-paint` and `first-contentful-paint` are of type PerformancePaintTiming  which also has duration always set to `0` (https://developer.mozilla.org/en-US/docs/Web/API/PerformancePaintTiming)

I have changed `(metric.startTime + metric.duration)` to just `metric.startTime`. Metrics are also defined statically in frontend. I've also moved converting `-` to `_` in metric names from backend to frontend.

aocenas reviewed (approved):
So this makes more sense to me, but maybe somebody who added this before could also take a look?

ifrost commented:
>  So this makes more sense to me, but maybe somebody who added this before could also take a look?

@torkelo - I've made some changes based on @aocenas comments. Could you please check if they make sense?
